drivers: crypto: Add initial support for rpi_pico sha256 accelerator

soburi opened the pull request:
Add basic support for RaspberryPi Pico's SHA256 hardware accelerator.

soburi committed:
drivers: crypto: Add initial support for rpi_pico sha256 accelerator

Add basic support for RaspberryPi Pico's SHA256 hardware accelerator.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: raspberrypi: rpi_pico2: Add sha256 accelerator

Added sha256 accelerator to rpi_pico.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: crypto: crypto_hash: Add rpi_pico2 to test target

Add rpi_pico2 to the crypto_hash test target.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
modules: hal_rpi_pico: Add `hardware_sha256` to include directories

Add `hardware_sha256/include` to `zephyr_include_directories`.
Use the header only.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
@ceolin @valeriosetti @ThreeEights 

Could you take a look if you have a bit of a while?

valeriosetti commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,202 @@
+
+static inline size_t bytes_to_write(struct rpi_pico_sha256_data *state, size_t data_size_bytes)
+{
+	return ((state->total_data_size + (SHA256_BLOCK_SIZE_BYTES - 1)) &
```
I was wondering if here we could use `ROUND_UP()` instead of this (and similarly on line 99). Your solution seems to be more performant since it uses bit-masks, whereas the one in `zephyr/include/zephyr/sys/util.h` might rely on compiler to optimize that computation. OTOH it would be more readable. Wdyt?

soburi replied:
Thank you for your good pointing out.
I got enough optimized code with ROUND_UP().
I applied your suggestion.

valeriosetti reviewed (commented):
The PR looks OK to me. I cannot test it locally because I don't have a board at hand, but at least the test that was extended builds correctly. I only left one minor non-blocking question/comment.

ThreeEights commented on drivers/crypto/CMakeLists.txt:
```diff
@@ -1,17 +1,22 @@
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_SMARTBOND            crypto_smartbond.c)
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_STM32                crypto_stm32.c)
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_TINYCRYPT_SHIM       crypto_tc_shim.c)
+# zephyr-keep-sorted-stop
```
Sorting - good idea!

ajf58 replied:
nit: I'd prefer it if the sorting is one commit, and the addition is a separate commit. The end result is the same, but it's easier to to see the two orthogonal things going on.

valeriosetti commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,199 @@
+
+static void write_padding(struct rpi_pico_sha256_data *state)
+{
+	/* Has to be a multiple of 64 bytes */
+
+	uint64_t size = ROUND_UP(state->total_data_size, SHA256_PADDING_DATA_BYTES);
```
```suggestion
	/* Has to be a multiple of 64 bytes */
	uint64_t size = ROUND_UP(state->total_data_size, SHA256_PADDING_DATA_BYTES);
```

I also have a question: the comment says that `size` must be multiple of 64 bytes, but `SHA256_PADDING_DATA_BYTES` is 9. Shouldn't we use `SHA256_BLOCK_SIZE_BYTES` instead here?

soburi replied:
This was a mistake when updating. Thanks for pointing that out.

valeriosetti commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,199 @@
+	update_internal(state, &one_bit, 1);
+
+	/* Zero unused padding */
+	add_zero_bytes(state, padding_size_bytes - SHA256_PADDING_DATA_BYTES);
```
Here we always call `add_zero_bytes` passing the difference between 2 `size_t` values. When `SHA256_PADDING_DATA_BYTES > padding_size_bytes` this results in a very large `size_t` which is then cast to `int32_t` inside `add_zero_bytes`. This looks a bit tricky to me; I think that adding something like:

```
if (padding_size_bytes > SHA256_PADDING_DATA_BYTES) {
    add_zero_bytes(state, padding_size_bytes - SHA256_PADDING_DATA_BYTES);
}
```

would simply the read and also the implementation inside `add_zero_bytes` since you don't need the cast to `int32_t` there. Wdyt?

soburi replied:
This code was originally from Pico-SDK, and is essentially for correctly counting updates every 4 bytes.

https://github.com/zephyrproject-rtos/zephyr/pull/85036/files#diff-60d26a5d8987cae9b3bad47c682ee0a4fca9df04aa5b5c3da3f53929b132c747R91

Here, a cast is performed so that the value is determined correctly even if it goes beyond 0 and becomes negative.

It is also possible to change the count logic so that it does not go beyond 0, like this: 

```
data_size_bytes -= MIN(4, data_size_bytes);
```

But in this case, the original process seems to be a more smart decision.

valeriosetti replied:
I'm not against the cast here, I'm against calling `add_zero_bytes` in any case no matter what value `padding_size_bytes` has.
Let me propose an example in order to clarify what I have in mind:

- let's say that we have a case where `padding_size_bytes < SHA256_PADDING_DATA_BYTES` so that `(padding_size_bytes - SHA256_PADDING_DATA_BYTES) < 0`
- in this case `write_padding` would call `add_zero_bytes` with `(int32_t) data_size_bytes < 0`
- once in `add_zero_bytes` the `while` loop would be completely skipped.

If instead we add:
```
if (padding_size_bytes > SHA256_PADDING_DATA_BYTES) {
    add_zero_bytes(state, padding_size_bytes - SHA256_PADDING_DATA_BYTES);
}
```
in `write_padding` as I proposed this problem would have been solved because `write_padding` would not have called `add_zero_bytes` at all.

valeriosetti replied:
It's OK to me now. Thanks!!!

valeriosetti reviewed (changes requested):
Thanks for addressing previous comment. However reviewing the PR again I found a couple of new ones that I missed before. Sorry

soburi commented:
@ceolin @ThreeEights @ajf58 

Could you take a look at it?

ajf58 commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,203 @@
+LOG_MODULE_REGISTER(sha_rpi_pico_sha256, CONFIG_CRYPTO_LOG_LEVEL);
+
+/*
+ * This is based on the pico_sha256 implementation in the Pico-SDK.
```
Original code is a https://github.com/raspberrypi/pico-sdk/blob/2.1.0/src/rp2_common/pico_sha256/sha256.c , it's 3-Clause BSD (`BSD-3-Clause`).

I feel like this file has lifted-and-shifted too much for someone to ignore (many lines of code are the same, just formatted to meet the Zephyr Project's coding style).

I'd be happy if this file had a BSD-3-Clause on it. Zephyr must be compatible with this, we link against BSD-3-Clause all over the place. It's not easy t osee examples of that, however.

Alternatively, we could modify the version of the source code in the `hal_rpi_pico`. This is something I prefer to avoid, but that feels better at ensuring the software licence is upheld.

soburi replied:
The codes derived from pico-sdk were separated into `crypto_rpi_pico_sha256.h`, and the license changed to BSD-3.

soburi replied:
> nit: I'd prefer it if the sorting is one commit, and the addition is a separate commit. The end result is the same, but it's easier to to see the two orthogonal things going on.

Addressed.

ajf58 reviewed (commented):
The change looks functionally good, but I am concerned about the outright copy and paste of BSD code, then relabeling it as Apache 2.

I feel like someone else needs to comment on this.

soburi commented:
This PR contains BSD-3 derived code. (drivers/crypto/crypto_rpi_pico_sha256.h)
I added the label because I wanted to get TSC approval.

ajf58 commented:
> This PR contains BSD-3 derived code. (drivers/crypto/crypto_rpi_pico_sha256.h) I added the label because I wanted to get TSC approval.

I think, based on the conversation at https://github.com/zephyrproject-rtos/zephyr/pull/81183#discussion_r2002693559 that this'll work .

My suggestion is
1.  Change (extend) https://github.com/zephyrproject-rtos/hal_rpi_pico/blob/zephyr/src/rp2_common/pico_sha256/sha256.c in a PR. This already is BSD-3-Clause, nothing would change there.
2. Update the west manifest for Zephyr to bring in the new HAL.
3. Make use of the new code from here.

I know that this may feel contradictory to an earlier stance I have of "don't change the Pico SDK unless you have to." I feel like this is one of those times.

That said, I also hope we can do this in a way that makes wiring up the DMA engine to the SHA256 accelerator easy: it's a prime example of where one wants to offload this whilst the CPUs do something more useful.

soburi commented:
@ajf58 

> That said, I also hope we can do this in a way that makes wiring up the DMA engine to the SHA256 accelerator easy: it's a prime example of where one wants to offload this whilst the CPUs do something more useful.

DMA processing must be based on Zephyr's API, and I have determined that this part cannot be used. This is the main reason why I forked this file.

ajf58 commented:
> @ajf58
> 
> > That said, I also hope we can do this in a way that makes wiring up the DMA engine to the SHA256 accelerator easy: it's a prime example of where one wants to offload this whilst the CPUs do something more useful.
> 
> DMA processing must be based on Zephyr's API, and I have determined that this part cannot be used. This is the main reason why I forked this file.

Yes, I understand what problem you found. The approach you took was to breach the terms of the licence, which is not cool. "fork"ing a file is not a thing.

My suggestion is written out above as 3 steps. If you don't understand the steps as written, please ask specific clarifying questions.

soburi commented:
@ajf58 

> Yes, I understand what problem you found. The approach you took was to breach the terms of the licence, which is not cool. "fork"ing a file is not a thing.

I've never heard of the BSD 3 clause license code can't forking.

ajf58 commented:
> @ajf58
> 
> > Yes, I understand what problem you found. The approach you took was to breach the terms of the licence, which is not cool. "fork"ing a file is not a thing.
> 
> I've never heard of the BSD 3 clause license code can't forking.

Your original approach was to:
1. Remove the original copyright notice. [Copyright (c) 2024 Raspberry Pi (Trading) Ltd.](https://github.com/raspberrypi/pico-sdk/blob/2.1.0/src/rp2_common/pico_sha256/sha256.c#L2)
2. Add your own name as the Copyright holder
3. Reissue the code under the Apache 2.0 license 

This, to me, is a concerning behaviour. It should not require other people to check/monitor/flag that you are breaching other licence terms, you should do this yourself.

After I flagged this you have subsequently corrected this by putting back the original copyright notice, and license. From a technical perspective I don't like how this has been achieved, but that is not the concern I am raising here.

soburi commented:
@ajf58 

> Your original approach was to:
> 
>     1. Remove the original copyright notice. [Copyright (c) 2024 Raspberry Pi (Trading) Ltd.](https://github.com/raspberrypi/pico-sdk/blob/2.1.0/src/rp2_common/pico_sha256/sha256.c#L2)
> 
>     2. Add your own name as the Copyright holder
> 
>     3. Reissue the code under the Apache 2.0 license
> 
> 
> This, to me, is a concerning behaviour. It should not require other people to check/monitor/flag that you are breaching other licence terms, you should do this yourself.
> 
> After I flagged this you have subsequently corrected this by putting back the original copyright notice, and license. From a technical perspective I don't like how this has been achieved, but that is not the concern I am raising here.

This is exactly as you pointed out.
I had mentioned that I was using it, but I had not properly addressed the license.
Thank you for pointing that out.

d3zd3z commented:
To be clear here. Changing an existing license declaration on a file can _only_ be done by permission of all of the contributing authors. Changing license declarations is _much_ worse than bringing in the files with the licenses on them (and figuring out what needs to happen to use the code under the license the authors released the code under).

The license lines in the file are an indication, from the author, of the terms that they allow their code to be used under them. Removing those declarations is in bad-faith, and is something we have to be very careful _never_ happens to code in this project.

nashif commented:
This was discussed in the TSC and it was agreed that we will not accept non original, Apache licensed code for this type of driver. Please use another method of integration.

soburi commented:
Thank you for your comment.

@d3zd3z @nashif 

I am aware that this is not a good situation, and I am very sorry.
I also thank @ajf58  for pointing this out to me.

I think there is another way to deal with this, so I will carefully reconsider it.

ajf58 commented:
> I think there is another way to deal with this, so I will carefully reconsider it.

Agreed. There have been a few PRs recently where we need them to stay Apache 2.0. I think the best way to do this is in the HAL, which is where several of them originate from. I don't think there's much of a convention across the different vendor HALs here, but I'll take a look.

soburi commented:
Add a public interface to the hal side and change the policy to use it.

valeriosetti commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,108 @@
+{
+	struct rpi_pico_sha256_data *data = dev->data;
+
+	assert(!data->state.locked);
```
Isn't `assert()` a bit too strong here? I suggest adding a mutex in `struct rpi_pico_sha256_data` and use it here:
- lock the mutex
- check `data->state.locked` and return if already set (releasing the mutex, of course before returning)
- set `data->state.locked` (not on line 79 as it is done now)
- unlock the mutex
- remember to unset `data->state.locked` in `rpi_pico_sha256_hash_session_free()`

In this way you can have 2 threads willing to use the hash driver at the same time and then the one that comes later will just wait for the resource to be free.

soburi commented:
@valeriosetti @ajf58 

Now that the 4.2 release is finished, can we resume?

valeriosetti commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,121 @@
+		return -EINVAL;
+	}
+
+	data->state.cache_used = 0;
+	data->state.cache.word = 0;
+	data->state.total_data_size = 0;
+
```
Is it OK to clear the `state` even if `finish == false` (multi step hash computation)?

valeriosetti replied:
> It seems like we should also consider modifying the tests.

While doing [another review](https://github.com/zephyrproject-rtos/zephyr/pull/93923/files#diff-8c410cc52e11be3967d86cc390ba80936868f85d9997c1c1288aebb71b0dc16aR60) yesterday I realized that perhaps not all drivers support multipart hash computation. So in the end what you did might be already OK for the time being (you can extend that later if you prefer); please only add a check like the one done in that PR and I think it should be fine.

dsseng commented on drivers/crypto/CMakeLists.txt:
```diff
@@ -1,20 +1,25 @@
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_ATAES132A            crypto_ataes132a.c)
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_CC23X0               crypto_cc23x0.c)
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_INTEL_SHA            crypto_intel_sha.c)
 zephyr_library_sources_ifdef(CONFIG_CRYPTO_IT51XXX_SHA		crypto_it51xxx_sha.c)
```
Despite all other entries being transitioned to uniform space indentation, this remains tabs

dsseng commented on drivers/crypto/CMakeLists.txt:
```diff
@@ -1,20 +1,25 @@
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_SI32                 crypto_si32.c)
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_SMARTBOND            crypto_smartbond.c)
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_STM32                crypto_stm32.c)
+zephyr_library_sources_ifdef(CONFIG_CRYPTO_STM32_HASH		crypto_stm32_hash.c)
```
Could this also be converted to spaces for alignment?

dsseng commented on drivers/crypto/Kconfig.rpi_pico:
```diff
@@ -0,0 +1,10 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config CRYPTO_RPI_PICO_SHA256
+	bool "RasberryPi Pico SHA256 Accelerator driver"
```
```suggestion
	bool "Raspberry Pi Pico SHA256 Accelerator driver"
```

or maybe even like this, as Pico is a board and the driver is for SoC. It's opinionated though, feel free to reject:

```suggestion
	bool "Raspberry Pi microcontrollers SHA256 Accelerator driver"
```

dsseng commented on drivers/crypto/Kconfig.rpi_pico:
```diff
@@ -0,0 +1,10 @@
+	depends on DT_HAS_RASPBERRYPI_PICO_SHA256_ENABLED
+	select PICOSDK_USE_SHA256
+	help
+	  RaspberryPi Pico SHA256 accelarator driver.
```
```suggestion
	  Raspberry Pi Pico SHA256 accelarator driver.
```

dsseng commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,135 @@
+
+static int crypto_rpi_pico_sha256_query_hw_caps(const struct device *dev)
+{
+	return (CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS);
```
personal opinion: readability would be better without parentheses:

```suggestion
	return CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS;
```

soburi replied:
Fixed.

dsseng commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,135 @@
+	sha256_set_bswap(true);
+	sha256_start();
+
+	pico_sha256_update(&data->state, pkt->in_buf, pkt->in_len);
```
Are there limits on the input size by the HAL? If so, please test for them to return EINVAL

soburi replied:
There are no particular restrictions, so it's not a problem.

ajf58 commented on tests/crypto/crypto_hash/src/main.c:
```diff
@@ -15,6 +15,8 @@
 #elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_hash)
 #define CRYPTO_DEV_COMPAT st_stm32_hash
+#elif DT_HAS_COMPAT_STATUS_OKAY(raspberrypi_pico_sha256)
+#define CRYPTO_DEV_COMPAT raspberrypi_pico_sha256
```
Nit: I'd prefer this in alphabetical order (MBED, raspberrypi, renesas, st). GitHub doesn't support multiline suggestions unfortunately.

soburi replied:
Fixed.

ajf58 reviewed (changes requested):
One suggestion that needs correcting, the other is a tiny nit. Otherwise it looks good.

kartben commented on dts/bindings/crypto/raspberrypi,pico-sha256.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2025 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+description: RaspberryPi Pico SHA256 accelerator
```
"SHA-256", even :)

```suggestion
description: Raspberry Pi Pico SHA-256 accelerator
```

soburi replied:
Fixed.

soburi commented:
@ajf58 @valeriosetti 

Please confirm the change request again.

soburi commented on tests/crypto/crypto_hash/socs/rp2350a_m33.overlay:
I haven't checked it yet, but if necessary I'll address it with an additional PR.

ajf58 replied:
I don't think this should be a blocker to this PR. If you haven't tested it, that's fine, and this comment in the PR is sufficient to document that. If you _have_ tested it and it doesn't work, that's fine too.

ceolin commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,135 @@
+	struct crypto_rpi_pico_sha256_data *data = dev->data;
+	k_spinlock_key_t key;
+
+	if (!data->state.locked) {
```
you have a TOCTOU here and ended up releasing it more than once. I don't know how much of this is a problem because the important lock here seems to be `BOOTROM_LOCK_SHA_256`. Btw can't you simply inquire whether or not this lock is LOCKED instead of having the spin lock + `state.locked` ?

soburi commented on drivers/crypto/crypto_rpi_pico_sha256.c:
```diff
@@ -0,0 +1,135 @@
+	k_spinlock_key_t key;
+	int ret;
+
+	if (data->state.locked) {
```
This has been corrected in conjunction with the above fix.

> and here you could possibly just bootrom_try_acquire_lock(BOOTROM_LOCK_SHA_256); and return EBUSY ...

Applied. Thank you.

soburi committed:
drivers: spi: spi_rpi_pico_pio: Fix pio_program variable constness

Since RPI_PICO_PIO_GET_PROGRAM returns const variable, so changed
`pio_program_t *` variables in data to `const pio_program_t *`.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

ajf58 commented on tests/crypto/crypto_hash/testcase.yaml:
```diff
@@ -3,6 +3,7 @@ tests:
     platform_allow:
       - native_sim
       - nucleo_u575zi_q
+      - rpi_pico2/rp2350a/m33
```
nit: (not a blocker): I imagine this should just work with the Hazard3 cores as well, since there's nothing specific to the Cortex M33s in your driver.
```suggestion
      - rpi_pico2/rp2350a/hazard3
      - rpi_pico2/rp2350a/m33
```
